File: paleo_workbench/ui/workstation/merge_features_dialog.py
"""无缝合并确认对话框（拓扑编辑迁移 M3 §4）。

预填面积最大要素属性（下拉可换源要素），相分类字段冲突高亮可改。
确认后由调用方把 ``result_payload()`` 交给桥 ``merge_mirror_features``。
轻量 QDialog（对照 ``ui/crs_guidance.py``）。
"""
from __future__ import annotations

from PySide6.QtWidgets import (
    QComboBox,
    QDialog,
    QDialogButtonBox,
    QFormLayout,
    QLabel,
    QLineEdit,
    QVBoxLayout,
)

from paleo_workbench.mapping.merge_attributes import plan_merge_attributes

__all__ = ["MergeFeaturesDialog"]


class MergeFeaturesDialog(QDialog):
    """合并确认：源要素下拉 + 可编辑属性（冲突高亮）。"""

    def __init__(self, records, *, facies_fields=("facies",), parent=None):
        super().__init__(parent)
        self.setObjectName("MergeFeaturesDialog")
        self.setWindowTitle("合并要素")
        self._records = [dict(record) for record in records]
        self.plan = plan_merge_attributes(self._records, facies_fields=facies_fields)
        self._attributes = dict(self.plan.get("attributes") or {})
        self._edits: dict[str, QLineEdit] = {}

        layout = QVBoxLayout(self)
        intro = QLabel("合并后保留一块几何。属性默认取面积最大的要素，可改源或改值。")
        intro.setWordWrap(True)
        layout.addWidget(intro)

        self._source = QComboBox(self)
        for record in self._records:
            feature_id = str(record.get("id") or "")
            self._source.addItem(feature_id, feature_id)
        target = str(self.plan.get("target_id") or "")
        index = self._source.findData(target)
        if index >= 0:
            self._source.setCurrentIndex(index)
        self._source.currentIndexChanged.connect(self._on_source_changed)
        layout.addWidget(QLabel("属性来源："))
        layout.addWidget(self._source)

        self._form = QFormLayout()
        layout.addLayout(self._form)
        self._rebuild_form()

        buttons = QDialogButtonBox(
            QDialogButtonBox.StandardButton.Ok | QDialogButtonBox.StandardButton.Cancel,
            parent=self)
        buttons.accepted.connect(self.accept)
        buttons.rejected.connect(self.reject)
        layout.addWidget(buttons)

    def _conflicts(self) -> set[str]:
        conflicts = self.plan.get("conflicts") or {}
        return {str(key) for key in conflicts}

    def _facies_fields(self) -> set[str]:
        return {str(name) for name in (self.plan.get("facies_fields") or ("facies",))}

    def _rebuild_form(self) -> None:
        while self._form.rowCount():
            self._form.removeRow(0)
        self._edits.clear()
        keys: list[str] = []
        for key in self._attributes:
            if key not in keys:
                keys.append(str(key))
        for key in sorted(self._conflicts()):
            if key not in keys:
                keys.append(key)
        for key in keys:
            edit = QLineEdit(self)
            value = self._attributes.get(key)
            edit.setText("" if value is None else str(value))
            edit.textChanged.connect(
                lambda text, field=key: self._attributes.__setitem__(field, text))
            if key in self._conflicts():
                # 冲突高亮：相分类字段更醒目。
                color = "#fde8a0" if key in self._facies_fields() else "#fff3cd"
                edit.setStyleSheet(f"QLineEdit {{ background: {color}; }}")
                edit.setToolTip("所选要素该字段值不一致")
            self._form.addRow(key, edit)
            self._edits[key] = edit

    def _on_source_changed(self, index: int) -> None:
        if index < 0 or index >= len(self._records):
            return
        record = self._records[index]
        self.plan["target_id"] = str(record.get("id") or "")
        properties = record.get("properties") if isinstance(record.get("properties"), dict) else {}
        self._attributes = dict(properties)
        self._rebuild_form()

    def set_field_value(self, field: str, value) -> None:
        """测试/调用方改值入口（同步到编辑框）。"""
        field = str(field)
        text = "" if value is None else str(value)
        self._attributes[field] = text
        edit = self._edits.get(field)
        if edit is not None:
            edit.setText(text)
        else:
            self._rebuild_form()

    def result_payload(self) -> dict[str, object]:
        """桥 merge_mirror_features 的 attrs_json 对象。"""
        return {
            "target_id": str(self.plan.get("target_id") or ""),
            "attributes": dict(self._attributes),
        }

File: libs/ui_composite/src/merge_features_dialog.cpp
#include <pwb/ui_composite/merge_features_dialog.hpp>

#include <pwb/ui_composite/merge_plan.hpp>

#include <QDialogButtonBox>
#include <QLabel>
#include <QVBoxLayout>

namespace pwb::ui_composite {

namespace {

// Python str(v) parity for JSON scalars; null -> "".
std::string json_text(const Json& value) {
    if (value.is_null()) return "";
    if (value.is_string()) return value.get<std::string>();
    if (value.is_boolean()) return value.get<bool>() ? "True" : "False";
    if (value.is_number()) {
        if (value.is_number_integer()) {
            return std::to_string(value.get<long long>());
        }
        if (value.is_number_unsigned()) {
            return std::to_string(value.get<unsigned long long>());
        }
        std::string text = value.dump();
        // Python prints floats without trailing ".0" only via repr —
        // Json dump already matches (e.g. 12.0 -> "12.0").
        return text;
    }
    return value.dump();
}

QString qstr(const std::string& text) {
    return QString::fromStdString(text);
}

}  // namespace

MergeFeaturesDialog::MergeFeaturesDialog(
    std::vector<Json> records, std::vector<std::string> facies_fields,
    QWidget* parent)
    : QDialog(parent),
      records_(std::move(records)),
      plan_(plan_merge_attributes(records_, facies_fields)),
      attributes_(plan_.value("attributes", Json::object())) {
    setObjectName("MergeFeaturesDialog");
    setWindowTitle(QStringLiteral("合并要素"));

    auto* layout = new QVBoxLayout(this);
    auto* intro = new QLabel(
        QStringLiteral("合并后保留一块几何。属性默认取面积最大的要素，可改源或改值。"),
        this);
    intro->setWordWrap(true);
    layout->addWidget(intro);

    source_ = new QComboBox(this);
    for (const Json& record : records_) {
        const std::string feature_id =
            json_text(record.value("id", Json()));
        source_->addItem(qstr(feature_id), qstr(feature_id));
    }
    const std::string target = plan_.value("target_id", "");
    const int index = source_->findData(qstr(target));
    if (index >= 0) source_->setCurrentIndex(index);
    connect(source_, &QComboBox::currentIndexChanged, this,
            &MergeFeaturesDialog::on_source_changed);
    layout->addWidget(new QLabel(QStringLiteral("属性来源："), this));
    layout->addWidget(source_);

    form_ = new QFormLayout();
    layout->addLayout(form_);
    rebuild_form();

    auto* buttons = new QDialogButtonBox(
        QDialogButtonBox::Ok | QDialogButtonBox::Cancel, this);
    connect(buttons, &QDialogButtonBox::accepted, this,
            &MergeFeaturesDialog::accept);
    connect(buttons, &QDialogButtonBox::rejected, this,
            &MergeFeaturesDialog::reject);
    layout->addWidget(buttons);
}

std::set<std::string> MergeFeaturesDialog::conflicts() const {
    std::set<std::string> out;
    const Json conflicts = plan_.value("conflicts", Json::object());
    if (conflicts.is_object()) {
        for (auto it = conflicts.begin(); it != conflicts.end(); ++it) {
            out.insert(it.key());
        }
    }
    return out;
}

std::set<std::string> MergeFeaturesDialog::facies_field_names() const {
    std::set<std::string> out;
    const Json fields = plan_.value("facies_fields", Json::array());
    if (fields.is_array()) {
        for (const Json& name : fields) {
            if (name.is_string()) out.insert(name.get<std::string>());
        }
    }
    if (out.empty()) out.insert("facies");
    return out;
}

void MergeFeaturesDialog::rebuild_form() {
    while (form_->rowCount() > 0) form_->removeRow(0);
    edits_.clear();
    std::vector<std::string> keys;
    if (attributes_.is_object()) {
        for (auto it = attributes_.begin(); it != attributes_.end(); ++it) {
            keys.push_back(it.key());
        }
    }
    for (const std::string& key : conflicts()) {
        if (std::find(keys.begin(), keys.end(), key) == keys.end()) {
            keys.push_back(key);
        }
    }
    const std::set<std::string> conflict_set = conflicts();
    const std::set<std::string> facies = facies_field_names();
    for (const std::string& key : keys) {
        auto* edit = new QLineEdit(this);
        const Json value =
            attributes_.is_object() && attributes_.contains(key)
                ? attributes_[key]
                : Json();
        edit->setText(qstr(json_text(value)));
        connect(edit, &QLineEdit::textChanged, this,
                [this, key](const QString& text) {
                    attributes_[key] = text.toStdString();
                });
        if (conflict_set.count(key)) {
            // 冲突高亮：相分类字段更醒目。
            const QString color = facies.count(key) ? "#fde8a0" : "#fff3cd";
            edit->setStyleSheet(
                QStringLiteral("QLineEdit { background: %1; }").arg(color));
            edit->setToolTip(QStringLiteral("所选要素该字段值不一致"));
        }
        form_->addRow(qstr(key), edit);
        edits_[key] = edit;
    }
}

void MergeFeaturesDialog::on_source_changed(int index) {
    if (index < 0 || index >= static_cast<int>(records_.size())) return;
    const Json& record = records_[static_cast<size_t>(index)];
    plan_["target_id"] = json_text(record.value("id", Json()));
    const Json properties = record.value("properties", Json::object());
    attributes_ = properties.is_object() ? properties : Json::object();
    rebuild_form();
}

void MergeFeaturesDialog::set_field_value(const std::string& field,
                                          const Json& value) {
    const std::string text = json_text(value);
    attributes_[field] = text;
    auto it = edits_.find(field);
    if (it != edits_.end() && it->second != nullptr) {
        it->second->setText(qstr(text));
    } else {
        rebuild_form();
    }
}

Json MergeFeaturesDialog::result_payload() const {
    return {{"target_id", plan_.value("target_id", "")},
            {"attributes", attributes_}};
}

}  // namespace pwb::ui_composite
